sdk/client/src/stream/streaming_client.rs

```diff
@@ -0,0 +1,204 @@
+        let msg = match self.stream.recv().await {
+            None => return Err(Error::connection_closed(None::<Error>)),
+            Some(msg) => msg,
+        };
+
+        let msg = match msg {
+            Ok(msg) => msg,
+            Err(e) => {
+                self.error_sender.send(Err(e)).await.ok();
+                return Ok(());
+            }
+        };
+
+        // If there is no ID, we can't route this to anywhere but the catchall '_error' channel
+        let id = match &msg.id {
+            Some(id) => id,
+            None => {
+                self.error_sender.send(Ok(msg)).await.ok();
+                return Ok(());
+            }
+        };
+        // Send the message to the respective channel
+        let id = id.clone();
+        match self.subscriptions.read().await.get(&id) {
+            // If we could not send the subscription, or the channel is closed, make sure to clean up the subscription
+            Some(sender) => match sender.send(Ok(msg)).await {
+                Err(e) => {
+                    self.unsubscribe(&id).await;
+                    Err(Error::connection_closed(Some(e)))
+                }
+                Ok(_) => Ok(()),
+            },
+            // No such subscription exists
+            None => {
+                self.error_sender
+                    .send(Err(Error::subscription_id_not_found(
+                        Some(msg),
+                        None::<Error>,
+                    )))
+                    .await
+                    .ok();
+                self.unsubscribe(&id).await;
+                Ok(())
+            }
+        }
+    }
+
+    pub async fn unsubscribe(&self, id: &Id) {
+        self.subscriptions.write().await.remove(id);
+    }
+}
+
+pub struct StreamingClient<C: StreamingClientTransport + 'static> {
+    client: Arc<RwLock<C>>,
+    subscriptions: Arc<RwLock<HashMap<Id, StreamingClientSender>>>,
+    channel_size: usize,
+    error_sender: StreamingClientSender,
+    channel_task: Option<JoinHandle<Result<()>>>,
+}
+
+impl<C: StreamingClientTransport + 'static> StreamingClient<C> {
+    pub async fn new(client: C, channel_size: usize) -> (Self, StreamingClientReceiver) {
+        let subscriptions = Arc::new(RwLock::new(HashMap::new()));
+        let (error_sender, error_receiver) =
+            mpsc::channel::<Result<StreamJsonRpcResponse>>(channel_size);
+
+        subscriptions
+            .write()
+            .await
+            .insert(Id::String(Box::from("_errors")), error_sender.clone());
+
+        let (stream, client) = client.get_stream();
+
+        let mut sct = Self {
+            client: Arc::new(RwLock::new(client)),
+            subscriptions,
+            channel_size,
+            error_sender,
+            channel_task: None,
+        };
+
+        sct.start_channel_task(stream);
+
+        (sct, error_receiver)
+    }
+
+    #[allow(unused)]
```

Comment:
why this internal func allow unused? missing "pub" keyword?

Reply:
yes, thank you

Reply:
Please don't have anything marked as unused or dead_code. Make it public if its intended to be used downstream, mark it `cfg(test)` if its only used for testing.